File: app/src/main/java/com/stathis/elmepaunivapp/recyclerviews/LessonsAdapter.java
package com.stathis.elmepaunivapp.recyclerviews;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.stathis.elmepaunivapp.R;
import com.stathis.elmepaunivapp.abstraction.DiffItemCallbackClass;
import com.stathis.elmepaunivapp.listeners.LessonClickListener;
import com.stathis.elmepaunivapp.ui.syllabus_lessons.model.Lesson;

import java.util.List;

public class LessonsAdapter extends ListAdapter<Lesson,LessonsViewHolder> {

    private LessonClickListener listener;

    public LessonsAdapter(LessonClickListener listener) {
        super(new DiffItemCallbackClass<Lesson>());
        this.listener = listener;
    }

    @NonNull
    @Override
    public LessonsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.lesson_item_row, parent, false);
        return new LessonsViewHolder(view,listener);
    }

    @Override
    public void onBindViewHolder(@NonNull LessonsViewHolder holder, int position) {
        holder.present(getItem(position));
    }

}

File: app/src/main/java/com/stathis/elmepaunivapp/ui/lessons/adapter/LessonsAdapter.kt
package com.stathis.elmepaunivapp.ui.lessons.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.ListAdapter
import com.stathis.elmepaunivapp.R
import com.stathis.elmepaunivapp.abstraction.DiffItemClass
import com.stathis.elmepaunivapp.callbacks.ElmepaClickListener
import com.stathis.elmepaunivapp.abstraction.LocalModel
import com.stathis.elmepaunivapp.databinding.HolderEmptyItemRowBinding
import com.stathis.elmepaunivapp.databinding.HolderLessonsHeaderItemBinding
import com.stathis.elmepaunivapp.databinding.LessonItemRowBinding
import com.stathis.elmepaunivapp.ui.lessons.model.LessonHeader
import com.stathis.elmepaunivapp.ui.lessons.model.Lesson

class LessonsAdapter(private val callback: ElmepaClickListener) : ListAdapter<LocalModel, LessonsViewHolder>(DiffItemClass<LocalModel>()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): LessonsViewHolder {
        val view = when(viewType){
            R.layout.holder_lessons_header_item -> HolderLessonsHeaderItemBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            R.layout.lesson_item_row -> LessonItemRowBinding.inflate(LayoutInflater.from(parent.context),parent,false)
            else -> HolderEmptyItemRowBinding.inflate(LayoutInflater.from(parent.context),parent,false)
        }
        return LessonsViewHolder(view,callback)
    }

    override fun onBindViewHolder(holder: LessonsViewHolder, position: Int) {
        holder.bindData(getItem(position))
    }

    override fun getItemViewType(position: Int): Int = when (getItem(position)) {
        is LessonHeader -> R.layout.holder_lessons_header_item
        is Lesson -> R.layout.lesson_item_row
        else -> R.layout.holder_empty_item_row
    }
}